pyccl/base/ccl_object.py

```diff
@@ -143,6 +143,24 @@ def Funlock(cls, cl, name, mutate: bool):
             newfunc = cls.unlock_instance(mutate=mutate)(func)
             setattr(cl, name, newfunc)
 
+    # TODO: Remove for CCLv3 & replace with Funlock.
+    @classmethod
+    def Funlock2(cls, cl, name):
+        """Replacement for Funlock incompatibility with signature.bind."""
+
+        def lock_on_exit(func):
+            @functools.wraps(func)
+            def wrapper(self, *args, **kwargs):
+                out = func(self, *args, **kwargs)
+                self._object_lock.lock()
+                return out
+            return wrapper
+
+        func = vars(cl).get(name)
+        if func is not None:
+            newfunc = lock_on_exit(func)
+            setattr(cl, name, newfunc)
+
```

Comment:
Some `__init__` methods are decorated with `@warn_api`. However, there's another decorator on top of that that is put there internally by `CCLObject`: `@unlock_instance`. So, the equivalent full code would be:
```python
@unlock_instance
@warn_api
def __init__(self, ...):
    ...
```

As per Python rules, decorators run from top to bottom (since they wrap the inside function), so `unlock_instance` is run first. With #1044 I changed the way we bind `argv` to argument names from the naive implementation we had, to a much better one which uses `signature.bind` (which automatically binds the input to the function's signature - so you don't have to play with indices).

Well, with that improvement, `signature.bind` complains when `MassFunc.__init__` and the like from the halo model get an extra argument (which used to be `cosmo` and has now been removed). `warn_api` preserves API for the extra `cosmo`, but since `unlock_instance` runs first, it beats it to it, and raises:
```python
ERROR pyccl/tests/test_pkhm.py - TypeError: too many positional arguments
ERROR pyccl/tests/test_tkk1h.py - TypeError: too many positional arguments
```

So, in order to keep preserving the API after the improvement with `signature.bind` I patched it so that `unlock_instance` doesn't wrap `__init__` any more. I simply created another decorator which locks the object once `__init__` has run.

All of this will be deleted for v3, it's just a patch for the final v2 release.